src/SeoPaneComponent.js

```diff
@@ -2,21 +2,45 @@ import React, {useState} from 'react'
 import PropTypes from 'prop-types'
 import {Text, Stack, Box, Card, Label, Flex, TabList, Tab, TabPanel, Spinner} from '@sanity/ui'
 import {useQuery} from 'react-query'
+import delve from 'dlv'
 import SerpPreview from 'react-serp-preview'
 
+import asyncCall from './lib/asyncCall'
 import performSeoReview from './lib/performSeoReview'
 import {renderRatingToColor} from './lib/renderRatingToColor'
 import {resultsLabels} from './lib/resultsLabels'
 
+import ErrorStack from './ErrorStack.js'
 import Feedback from './Feedback'
 
-export default function SeoPaneComponent({revision, url, keywords, synonyms}) {
+export default function SeoPaneComponent({document, options}) {
   const [tab, setTab] = useState('')
 
   // The `revision` key updates when the document does, refreshing the query
   const {data, isLoading, error} = useQuery(
-    [`seoReview`, revision],
-    async () => performSeoReview(url, keywords, synonyms),
+    [`seoReview`, document._rev],
+    async () => {
+      if (!document._id) throw new Error('Document is not published')
+      else if (!options.keywords) badOption('keywords')
+      else if (!options.url) badOption('url')
+
+      let [keywords, synonyms, url] = await Promise.all([
+        asyncCall(options.keywords, document),
+        asyncCall(options.synonyms, document),
+        asyncCall(options.url, document),
+      ])
+
+      // Visits document path when strings because the asyncCall will have same value as options
+      if (keywords && keywords === options.keywords) keywords = delve(keywords, document)
+      if (synonyms && synonyms === options.synonyms) synonyms = delve(synonyms, document)
```

Comment:
```suggestion
      if (keywords && keywords === options.keywords) keywords = delve(document, keywords)
      if (synonyms && synonyms === options.synonyms) synonyms = delve(document, synonyms)
```

This library's params go `obj` first, `keypath` second. 

Everything else looks good to go :)

Reply:
Ah good catch, I must have obliterated the lines and recreated them from memory, although a bit backwards.